dali/kernels/common/fast_hash.h

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_FAST_HASH_H_
+#define DALI_KERNELS_COMMON_FAST_HASH_H_
+
+#include <cstdint>
+#include <cstddef>
+#include <functional>
+#include "dali/core/force_inline.h"
+
+namespace dali {
+namespace kernels {
+
+struct fast_hash_t {
+  uint32_t data[8];
+
+  bool operator==(const fast_hash_t &other) const {
+    for (int i = 0; i < 8; i++) {
+      if (data[i] != other.data[i])
+        return false;
+    }
+    return true;
+  }
+  bool operator!=(const fast_hash_t &other) const {
+    return !(*this == other);
+  }
+  bool operator<(const fast_hash_t &other) const {
+    for (int i = 0; i < 8; i++) {
+      if (data[i] < other.data[i])
+        return true;
+      if (data[i] > other.data[i])
+        return false;
+    }
+    return false;
+  }
+};
+
+DALI_FORCEINLINE
+constexpr uint32_t rotl(uint32_t x, uint8_t r) {
+  return (x << r) | (x >> (32-r));
+}
+
+inline void fast_hash(fast_hash_t &hash, const void *data, size_t n) {
```

Comment:
Can you provide any reference where this formula came from?

Reply:
It's more like a general knowledge. It's based on xxHash 3, but uses circular propagation to avoid accumulating all entropy in just one word when the data has power-of-2 periodic pattern whwere only one byte/word changes.

Reply:
Could you add this as a comment?